### ISSUE TITLE
Unable to start modules on poor internet due to Client.Timeout exceeded while awaiting headers

### ISSUE THREAD

**Issue by mathiash98** [Azure/iotedge#3069]
## Expected Behavior
EdgeAgent should not require stable connection to remote repository to start module when docker image already is downloaded.

## Current Behavior
EdgeAgent will now check remote repository for docker image before starting module even though docker image already is downloaded.

## Steps to Reproduce
1. Have an edgeDevice with bad enough internet connection that timeout occurs.
2. Confirm that docker image already is downloaded
3. Trigger a restart of edgeAgent (either by new version in manifest or manual restart)
4. Observe that edgeAgent is producing timeout errors when starting modules due to timout exceptions when checking local docker image with remote. This leads to not starting module.

## Context (Environment)
We deployed a new deployment manifest which had a new edgeAgent version. This lead to edgeAgent being restarted.
Before we deployed, we manually pulled all required images for reasons specified in #3053 
After edgeAgent was restarted, all modules stops (naturally). But because of poor connection we experienced downtime of 30 minutes because docker was unable to check that the local image was correct to remote repository. (Unsure of edgeAgent's behavior but it is performing some connection to internet which fails.)

</details>

### Device Information
* Host OS [e.g. Ubuntu 16.04, Ubuntu 18.04, Windows IoT Core]: Ubuntu 18.04
* Architecture [e.g. amd64, arm32, arm64]:  amd64
* Container OS [e.g. Linux containers, Windows containers]: Linux

### Runtime Versions
* iotedged [run `iotedge version`]: 1.0.9.2
* Edge Agent [image tag (e.g. 1.0.0)]: 1.0.9.2
* Edge Hub [image tag (e.g. 1.0.0)]: 1.0.9.2

## Logs
<details>
<summary>iotedged logs</summary>

```
Jun 10 10:41:53 dev-kit4-tsbh03026955 iotedged[1787]: 2020-06-10T10:41:53Z [INFO] - Pulling image srtprodacr.azurecr.io/inrangeonpremapi:3309...
Jun 10 10:41:57 dev-kit4-tsbh03026955 iotedged[1787]: 2020-06-10T10:41:57Z [INFO] - Checking edge runtime status
Jun 10 10:41:57 dev-kit4-tsbh03026955 iotedged[1787]: 2020-06-10T10:41:57Z [INFO] - Edge runtime is running.
Jun 10 10:42:08 dev-kit4-tsbh03026955 iotedged[1787]: 2020-06-10T10:42:08Z [WARN] - Could not pull image srtprodacr.azurecr.io/inrangeonpremapi:3309
Jun 10 10:42:08 dev-kit4-tsbh03026955 iotedged[1787]: 2020-06-10T10:42:08Z [WARN] -         caused by: Get https://srtprodacr.azurecr.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
```
</details>

<details>
<summary>edge-agent logs</summary>

```
<3> 2020-06-10 10:42:55.138 +00:00 [ERR] - Executing command for operation ["Command Group: (\n  [Command Group: (\n  [Prepare update module api]\n  [Stop module api]\n  [Update module api]\n)]\n  [Start module api]\n)"] failed.
Microsoft.Azure.Devices.Edge.Agent.Edgelet.EdgeletCommunicationException- Message:Error calling prepare update for module module api: Could not prepare update for module "api"
        caused by: Could not pull image srtprodacr.azurecr.io/inrangeonpremapi:3309
        caused by: Get https://srtprodacr.azurecr.io/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers), StatusCode:500, at:   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2019_11_05.ModuleManagementHttpClient.HandleException(Exception exception, String operation) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2019_11_05/ModuleManagementHttpClient.cs:line 210
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Versioning.ModuleManagementHttpClientVersioned.Execute[T](Func`1 func, String operation) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/versioning/ModuleManagementHttpClientVersioned.cs:line 127
   at Microsoft.Azure.Devices.Edge.Agent.Edgelet.Version_2019_11_05.ModuleManagementHttpClient.PrepareUpdateAsync(ModuleSpec moduleSpec) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Edgelet/version_2019_11_05/ModuleManagementHttpClient.cs:line 201
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 60
   at Microsoft.Azure.Devices.Edge.Agent.Core.Commands.GroupCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/commands/GroupCommand.cs:line 35
   at Microsoft.Azure.Devices.Edge.Agent.Core.LoggingCommandFactory.LoggingCommand.ExecuteAsync(CancellationToken token) in /home/vsts/work/1/s/edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Core/LoggingCommandFactory.cs:line 60
<3> 2020-06-10 10:42:55.139 +00:00 [ERR] - Step failed in deployment 180, continuing execution. Failure when running command Command Group: (
  [Command Group: (
  [Prepare update module api]
  [Stop module api]
  [Update module api]
)]
  [Start module api]
). Will retry in 00s.
```
</details>

## Additional Information
See #3053 for related issue with a somewhat suggestion for solution.

Probably similar issue as in #2892

**Comment by richma-ms** [comment]
Was the imagePullPolicy property set to "never" on edgeAgent?  If so it shouldn't try to reach out to the registry.

When the runtime gets updated, there could be changes across version where edgeAgent decides it's necessary to redeploy a module via an update, and update is implemented as remove -> create.  When the module gets re-created, the imagePullPolicy behavior should kick in.

**Comment by mathiash98** [comment]
We have never used the `imagePullPolicy` before, would we benefit from using it?

To be clear, this issue is not specific to edgeAgent or edgeHub module. The described scenario was when updating custom modules. Which I already had downloaded manually with `docker pull`

**Comment by richma-ms** [comment]
In this case it should.  Due to the Edge runtime update, EdgeAgent removed the old module and tried to redeploy it as if it's new. 
 At this point the pull command is originated from EdgeAgent, which is affected by the imagePullPolicy.

**Comment by richma-ms** [comment]
Closing due to inactivity, if you still have problems please open a new issue.
